qiskit_experiments/calibration_management/save_utils.py

```diff
@@ -177,13 +187,26 @@ def calibrations_to_dict(
             sched_obj.metadata.update(qubit_metadata)
         sched_entries.append(sched_obj)
 
+    max_qubit = max(
+        (max(k.qubits or (0,)) for k in cals._parameter_map if k.schedule is None),
+        default=0,
+    )
+    schedule_free_parameters = QuantumCircuit(max_qubit + 1)
```

Comment:
Seems like this is a hack. Qpy also saves instruction and QuantumCircuit metadata, and using QuantumCircuit format will increase output data size and may cause edge case in future Qpy module updates. Although QuantumCircuit provides effective representation of parameters tied to particular qubits, I would prefer directly saving the parameter object as you wrote as a second approach.

Reply:
> Seems like this is a hack.

Yes, it is a hack, but there is no supported way to directly save a `Parameter` object, so I think any method will be a bit of a hack.

> Qpy also saves instruction and QuantumCircuit metadata, and using QuantumCircuit format will increase output data size

In practice, I don't think the overhead is too large for qpy since it tries to be efficient and no extra metadata is being added to the circuit or instructions in this code. I was curious what the comparison really was so I did a test. Here is the code that I used:

```python
import json
import zlib
from io import BytesIO

from qiskit import QuantumCircuit
from qiskit.circuit import Instruction, Parameter
from qiskit.qpy import load, dump

from qiskit_experiments.framework.json import ExperimentEncoder


num_params = 100

qc = QuantumCircuit(1)
for idx in range(num_params):
    qc.append(Instruction("drive_freq", 0, 0, [Parameter(f"param{idx}")]))


param_list = []
for idx in range(num_params):
    param = Parameter("a")
    param_list.append({"qubits": [0], "parameter_name": f"param{idx}", "uuid": param._uuid.bytes})

qpy_file = BytesIO()
pos = dump(qc, qpy_file)
qpy_file.seek(0)
new_circuit = load(qpy_file)[0]
qpy_file.seek(0)


json_bytes = json.dumps(param_list, cls=ExperimentEncoder).encode("utf-8")
json_bytes_zip = zlib.compress(json_bytes)

qpy_bytes = qpy_file.read()
print("number of parameters: ", num_params)
print("qpy bytes: ", len(qpy_bytes))
print("zipped qpy bytes: ", len(zlib.compress(qpy_bytes)))
print("json bytes: ", len(json_bytes))
print("zipped json bytes: ", len(json_bytes_zip))
```

and the results I got were:

```
number of parameters:  100
qpy bytes:  7859                                                                                                       
zipped qpy bytes:  2214                       
json bytes:  16574                                                                                                     
zipped json bytes:  2954
```

and

```
number of parameters:  200                                                                                             
qpy bytes:  15659                                                                                                      
zipped qpy bytes:  4300                                  
json bytes:  33374
zipped json bytes:  5790
```

So without compression, the qpy is about half as big as a JSON output. With compression, they are closer though qpy is still smaller.

> may cause edge case in future Qpy module updates

My concern was that not using qpy would be more vulnerable to a future Qiskit update. Qiskit tries to keep backwards compatibility with qpy, maintaining loading of old qpy files in newer versions of Qiskit. That is more guarantee than we have that saving `{"name": param._name, "uuid": param._uuid.bytes.encode("utf-8")}` will always robustly serialize a `Parameter`.

> Although QuantumCircuit provides effective representation of parameters tied to particular qubits, I would prefer directly saving the parameter object as you wrote as a second approach.

So would you prefer the approach I used in my test code, like the following?

```python
schedule_free_parameters = [
    {"qubits": list(s.qubits), "parameter_name": p._name, "uuid": p._uuid.bytes}
    for s, p in cals._parameter_map.items()
    if s.schedule is None
]
```

Reply:
Our Json Encoder does serialize Parameter object :) https://github.com/Qiskit-Extensions/qiskit-experiments/blob/0ee488989b67d0239597741bf2bb3e51b5d68965/qiskit_experiments/framework/json.py#L516-L521

Note that `Parameter` is a subclass of `ParameterExpression`. Since this uses the protected function (this code is also used by IBM's provider), the API stability is not really guaranteed, but at least data structure must be preserved in future version through QPY.

Reply:
I wonder what the use case was for adding `ParameterExpression` to the encoder. In any case, it does not work currently:

![image](https://github.com/Qiskit-Extensions/qiskit-experiments/assets/1178780/85e86f95-d1f9-4921-a142-d56cc059b36a)

and this is the kind of thing that I would like to avoid by using a public interface. So what is your preference for how to save the parameters. The options I see are:

1. Save them in `QuantumCircuit`
2. Save the name and uuid directly
3. Fix the serializer's use of `_write_parameter_expression`. Note that I think is kind of difficult. I looked into it and I needed to change `_read_parameter_expression` to `_read_parameter_expression_v3` which takes two extra arguments. That would only read new parameters. Old ones would need the old `_read_parameter_expression` but the logic to choose between the two is in other qpy code. Also, note that this serializes a `Parameter` into a `ParameterExpression` so the `save_utils` code would need to pull the `Parameter` out of the `ParameterExpression` which starts to feel like pulling the `Parameter` out of the quantum circuit instructions.

Reply:
I checked how IBM Runtime does this serialization, and found that they implemented [a custom logic](https://github.com/Qiskit/qiskit-ibm-runtime/blob/6c2cf0e71b1273923086845d4383dc00cfa559f6/qiskit_ibm_runtime/utils/json.py#L273-L283
) in their JSON decoder. This means current implementation of the ExperimentDecoder is not correct. I agree using the protected method increases maintenance cost like this, and I'm fine with using a circuit as a container, considering our bandwidth for maintenance.

Reply:
Okay, sounds good. Did you have fairly equal feeling about serializing quantum circuit versus name and uuid? I had doubts about using qpy serialization on Parameter directly but I was uncertain about the other two options. I made uuid a public property of Parameter in Qiskit if we wanted to go with serializing name and uuid directly.